# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmIonCompile.cpp
# Commit: 349728105c40
# Full Hash: 349728105c40993896170f4c2f43e152a3466706
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2022-01-08 09:40:24
# Description:
#   Bug 1748602 - Backout bug 1642412. r=lth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D135346
# ==============================================================================

diff -r 32873022d2ed -r 349728105c40 js/src/wasm/WasmIonCompile.cpp
--- a/js/src/wasm/WasmIonCompile.cpp	Fri Jan 07 20:35:15 2022 +0000
+++ b/js/src/wasm/WasmIonCompile.cpp	Fri Jan 07 20:43:16 2022 +0000
@@ -1315,9 +1315,8 @@
       actualBase = extended;
     }
 
-    auto* ins =
-        MWasmBoundsCheck::New(alloc(), actualBase, boundsCheckLimit,
-                              bytecodeOffset(), MWasmBoundsCheck::Memory);
+    auto* ins = MWasmBoundsCheck::New(alloc(), actualBase, boundsCheckLimit,
+                                      bytecodeOffset());
     curBlock_->add(ins);
     actualBase = ins;
 
@@ -1702,19 +1701,6 @@
 
   /************************************************ Global variable accesses */
 
-  bool postBarrierFiltering(MDefinition* valueAddr, uint32_t lineOrBytecode) {
-    const SymbolicAddressSignature& callee = SASigPostBarrierFiltering;
-    CallCompileState args;
-    if (!passInstance(callee.argTypes[0], &args)) {
-      return false;
-    }
-    if (!passArg(valueAddr, callee.argTypes[1], &args)) {
-      return false;
-    }
-    finishCall(&args);
-    return builtinInstanceMethodCall(callee, lineOrBytecode, args);
-  }
-
   MDefinition* loadGlobalVar(unsigned globalDataOffset, bool isConst,
                              bool isIndirect, MIRType type) {
     if (inDeadCode()) {
@@ -1784,69 +1770,6 @@
     return valueAddr;
   }
 
-  MDefinition* loadTableField(const TableDesc& table, unsigned fieldOffset,
-                              MIRType type) {
-    uint32_t globalDataOffset =
-        offsetof(TlsData, globalArea) + table.globalDataOffset + fieldOffset;
-    auto* load = MWasmLoadTls::New(alloc(), tlsPointer_, globalDataOffset, type,
-                                   AliasSet::Load(AliasSet::WasmTableMeta));
-    curBlock_->add(load);
-    return load;
-  }
-  MDefinition* loadTableLength(const TableDesc& table) {
-    return loadTableField(table, offsetof(TableTls, length), MIRType::Int32);
-  }
-  MDefinition* loadTableElements(const TableDesc& table) {
-    return loadTableField(table, offsetof(TableTls, elements),
-                          MIRType::Pointer);
-  }
-
-  MDefinition* tableGetAnyRef(const TableDesc& table, MDefinition* index) {
-    // Load the table length and perform a bounds check with spectre index
-    // masking
-    auto* length = loadTableLength(table);
-    auto* check = MWasmBoundsCheck::New(
-        alloc(), index, length, bytecodeOffset(), MWasmBoundsCheck::Table);
-    curBlock_->add(check);
-    if (JitOptions.spectreIndexMasking) {
-      index = check;
-    }
-
-    // Load the table elements and load the element
-    auto* elements = loadTableElements(table);
-    auto* element = MWasmLoadTableElement::New(alloc(), elements, index);
-    curBlock_->add(element);
-    return element;
-  }
-  MDefinition* tableSetAnyRef(const TableDesc& table, MDefinition* index,
-                              MDefinition* value, uint32_t lineOrBytecode) {
-    // Load the table length and perform a bounds check with spectre index
-    // masking
-    auto* length = loadTableLength(table);
-    auto* check = MWasmBoundsCheck::New(
-        alloc(), index, length, bytecodeOffset(), MWasmBoundsCheck::Table);
-    curBlock_->add(check);
-    if (JitOptions.spectreIndexMasking) {
-      index = check;
-    }
-
-    // Load the table elements and compute the value's location
-    auto* elements = loadTableElements(table);
-    auto* loc =
-        MWasmDerivedIndexPointer::New(alloc(), elements, index, ScalePointer);
-    curBlock_->add(loc);
-
-    // Store the new value and then perform the post barrier
-    auto* store = MWasmStoreRef::New(alloc(), tlsPointer_, loc, value,
-                                     AliasSet::WasmTableElement);
-    curBlock_->add(store);
-    if (!postBarrierFiltering(loc, lineOrBytecode)) {
-      return nullptr;
-    }
-
-    return store;
-  }
-
   void addInterruptCheck() {
     if (inDeadCode()) {
       return;
@@ -4179,8 +4102,19 @@
   // the nursery, and the value stored will very frequently be in the nursery.
   // The C++ postbarrier performs any necessary filtering.
 
-  if (barrierAddr && !f.postBarrierFiltering(barrierAddr, lineOrBytecode)) {
-    return false;
+  if (barrierAddr) {
+    const SymbolicAddressSignature& callee = SASigPostBarrierFiltering;
+    CallCompileState args;
+    if (!f.passInstance(callee.argTypes[0], &args)) {
+      return false;
+    }
+    if (!f.passArg(barrierAddr, callee.argTypes[1], &args)) {
+      return false;
+    }
+    f.finishCall(&args);
+    if (!f.builtinInstanceMethodCall(callee, lineOrBytecode, args)) {
+      return false;
+    }
   }
 
   return true;
@@ -5456,19 +5390,9 @@
     return true;
   }
 
-  const TableDesc& table = f.moduleEnv().tables[tableIndex];
-  if (table.elemType.tableRepr() == TableRepr::Ref) {
-    MDefinition* ret = f.tableGetAnyRef(table, index);
-    if (!ret) {
-      return false;
-    }
-    f.iter().setResult(ret);
-    return true;
-  }
-
   uint32_t lineOrBytecode = f.readCallSiteLineOrBytecode();
 
-  const SymbolicAddressSignature& callee = SASigTableGetFunc;
+  const SymbolicAddressSignature& callee = SASigTableGet;
   CallCompileState args;
   if (!f.passInstance(callee.argTypes[0], &args)) {
     return false;
@@ -5566,16 +5490,7 @@
 
   uint32_t lineOrBytecode = f.readCallSiteLineOrBytecode();
 
-  const TableDesc& table = f.moduleEnv().tables[tableIndex];
-  if (table.elemType.tableRepr() == TableRepr::Ref) {
-    MDefinition* ret = f.tableSetAnyRef(table, index, value, lineOrBytecode);
-    if (!ret) {
-      return false;
-    }
-    return true;
-  }
-
-  const SymbolicAddressSignature& callee = SASigTableSetFunc;
+  const SymbolicAddressSignature& callee = SASigTableSet;
   CallCompileState args;
   if (!f.passInstance(callee.argTypes[0], &args)) {
     return false;
@@ -5615,14 +5530,33 @@
     return true;
   }
 
-  const TableDesc& table = f.moduleEnv().tables[tableIndex];
-
-  MDefinition* length = f.loadTableLength(table);
-  if (!length) {
-    return false;
-  }
-
-  f.iter().setResult(length);
+  uint32_t lineOrBytecode = f.readCallSiteLineOrBytecode();
+
+  const SymbolicAddressSignature& callee = SASigTableSize;
+  CallCompileState args;
+  if (!f.passInstance(callee.argTypes[0], &args)) {
+    return false;
+  }
+
+  MDefinition* tableIndexArg =
+      f.constant(Int32Value(tableIndex), MIRType::Int32);
+  if (!tableIndexArg) {
+    return false;
+  }
+  if (!f.passArg(tableIndexArg, callee.argTypes[1], &args)) {
+    return false;
+  }
+
+  if (!f.finishCall(&args)) {
+    return false;
+  }
+
+  MDefinition* ret;
+  if (!f.builtinInstanceMethodCall(callee, lineOrBytecode, args, &ret)) {
+    return false;
+  }
+
+  f.iter().setResult(ret);
   return true;
 }
 